src/lib.rs

```diff
@@ -220,8 +232,73 @@ pub trait Arbitrary<'a>: Sized {
     /// create a better implementation. If you are writing an `Arbitrary`
     /// implementation by hand, and your type can be part of a dynamically sized
     /// collection (such as `Vec`), you are strongly encouraged to override this
-    /// default with a better implementation. The
-    /// [`size_hint`] module will help with this task.
+    /// default with a better implementation, and also override [`faillible_size_hint`]
+    ///
+    /// ## How to implement this
+    ///
+    /// If the size hint calculation is a trivial constant and does not recurse into any other `size_hint` call, you should implement it in `size_hint`:
```

Comment:
Let's not muddy the waters by suggesting that sometimes people use the old size hint method and sometimes the new one.

Let's deprecate the old one and always recommend using the new one.

Reply:
See https://github.com/rust-fuzz/arbitrary/pull/185#discussion_r1759119669

We could say that this is "just" an optimization and that it's ok to break it with a minor release.
Arguably a lot of users use the derive macro or don't implement size_hint, so it could be justified. And marking it as deprecated would make this pretty visible to downstream.

I also don't like the duplicate API.

Reply:
The other comment about defaults and backwards compat makes sense, thanks for pointing that out.

That said, we should suggest that everyone move over to `try_size_hint` and ignore `size_hint` and/or implement it with `try_size_hint().unwrap_or_default()` and then avoid mentioning the deprecated method again.

Reply:
One of the things with my approach is that it does not require dependents to change what they call.

For example `libfuzzer_sys` does not care about the size hint being fallible, so it would just have been able `size_hint` and let the implementation deal with recursion internally.

I think the ideal setup would be to only have `try_size_hint` be part of the trait, and have an infallible `size_hint` in an extension trait (so users are not given the opportunity to override it) that does not even take a depth parameter for use in `libfuzzer_sys` for example. But that seems like adding more complexity and bring in naming conflicts.

Reply:
The backward compatibility issues will also mean that users will need to implement `size_hint` to call `try_size_hint` instead, and have to silence the `deprecated` warning to do so.

I think the issue this fixes most likely only affects users of the `derive` macros, and therefore I wanted to minimize the impact on other users. Especially given that a lot of types that can implement `arbitrary` will likely implement it on single, non recursive types and therefore they could still implement only `size_hint`.

Also, it's still possible to add deprecation attributes in the future if this does not happen to be the case.